### DISCUSSION TITLE
Error message: could not find function "ojs_define"

### DISCUSSION THREAD

**Discussion by ProcessMinerUK** [quarto-dev/quarto-cli#683] [Q&A] [answered]
I have just started using Quarto, but keep getting the error message - **could not find function "ojs_define"** when trying to define a value in an {r} code cell.

Quarto version: v0.9.276

**Comment by cscheid** [comment, marked as answer]
Hi - sorry you're running into trouble.

I'm having a hard time reproducing your example. Can you please take the following code:

    ---
    title: Hello
    server:
      type: shiny
    ---
    
    ```{ojs}
    //| panel: sidebar
    vars = ["Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width"]
    viewof xcol = Inputs.select(vars, {label: "X Variable"})
    viewof ycol = Inputs.select(vars, {label: "Y Variable", value: vars[1]})
    viewof count = Inputs.range([1, 9], {label: "Cluster Count", step: 1, value: 3})
    ```
    
    ```{ojs}
    //| panel: fill
    Plot.plot({
      color: {
        type: "ordinal",
        scheme: "category10"
      },
      marks: [
        Plot.dot(transpose(selectedData), {
          x: xcol,
          y: ycol,
          fill: (d, i) => clusters.cluster[i],
        }),
        Plot.dot(clusters.centers, { 
          x: d => d[0],
          y: d => d[1],
          r: 10,
          stroke: "black",
          fill: (d, i) => i + 1
        }),
      ]
    })
    ```
    
    ```{r}
    #| context: server
    
    selectedData <- reactive({
      iris[, c(input$xcol, input$ycol)]
    })
    
    clusters <- reactive({
      kmeans(selectedData(), input$count)
    })
    
    ojs_define(selectedData, clusters)
    ```

save it as `test-shiny.qmd` and then run `quarto serve test-shiny.qmd` on your shell? Thank you!

**Reply by cscheid** [reply]
I realize now that it's possible you weren't running a `shiny` server. In that case, this should be a minimal example, which works in my local environment:

    ---
    title: ojs_define test
    format: html
    ---
    
    ```{ojs}
    x
    ```
    
    ```{r}
    x = c(1,2,3,4)
    ojs_define(x)
    ```

This produces the following output for me:

<img width="862" alt="image" src="https://user-images.githubusercontent.com/285675/164038518-8e4ee58c-ed57-4c3c-90b0-5fe381ae8d8f.png">

Can you describe your installation and environment in more detail? 

- What OS are you running? 
- How did you attempt to generate the failing example?

**Reply by ProcessMinerUK** [reply]
Thank you for the example, I have now got this working.

**Reply by Doshin108** [reply]
I was expecting the function to work in R but it doesn't, it work on Render.  ::face-palm::   @cscheid your example tie the missing understanding together for me.   Appreciate you!
